src/Data/Text/Markdown.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Data.Text.Markdown where
+
+import CMarkGFM qualified as CMark
+import Control.Applicative ((<|>))
+import Data.Function ((&))
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Yaml
+import Swarm.Language.Parse (readTerm)
+import Swarm.Language.Pretty (prettyText)
+import Swarm.Language.Syntax (Syntax)
+import Swarm.Language.Pipeline (processParsedTerm, ProcessedTerm (..), prettyTypeErr)
+import Swarm.Language.Module (moduleAST)
+import Control.Monad (void)
+
+data Node
+  = Node NodeType [Node]
+  | LeafText Text
+  | LeafCode Syntax
+  | LeafCodeBlock String Syntax
+  deriving (Eq, Show)
+
+data NodeType
+  = Document
+  | Paragraph
+  deriving (Eq, Show)
+
+instance ToJSON Node where
+  toJSON = String . toText
+
+instance FromJSON Node where
+  parseJSON v =
+    withArray "markdown paragraphs" (foldMap parseJSON) v
+      <|> withText "markdown text" fromTextM v
+instance Semigroup Node where
+  Node Document cs1 <> Node Document cs2 = Node Document (cs1 <> cs2)
+  Node Document cs1 <> n2 = Node Document (cs1 <> [n2])
+  n1 <> Node Document cs2 = Node Document (n1 : cs2)
+  n1 <> n2 = Node Document [n1, n2]
+
+instance Monoid Node where
+  mempty = Node Document []
+
+fromText :: Text -> Node
+fromText = either error id . fromTextE
+
+fromTextM :: MonadFail m => Text -> m Node
+fromTextM = either fail pure . fromTextE
+
+fromTextE :: Text -> Either String Node
+fromTextE = fromGeneralNode . CMark.commonmarkToNode [CMark.optSourcePos] []
+
+-- TODO: use transformer kind of Error? (report all errors)
```

Comment:
@byorgey I noticed we do not depend on `transformers`, do you know of any type like [`Control.Applicative.Lift.Errors`](https://hackage.haskell.org/package/transformers-0.6.1.0/docs/Control-Applicative-Lift.html#t:Errors) that we could use?

It would be neat to show more errors at once where possible. :slightly_smiling_face:

Reply:
I vote to add `transformers` to our dependencies.  We depend on `mtl` which has a direct dependence on `transformers` anyway.

Reply:
Agreed, there's no reason we can't add a direct dependency on `transformers`.

Reply:
Thanks, @kostmo, I will try that. :+1:

Reply:
Actually, I may not want to do this here after all. Whenever I change code here, it impacts a lot of scenarios and they are reported per file already. At the end, Aeson does the parsing and error reporting, so there is not much I can do.